sources/api/netdog/src/cli/write_resolv_conf.rs

```diff
@@ -1,28 +1,177 @@
-        .context(error::NetworkDInterfaceStatusSnafu)?;
-    let dns_settings = DnsSettings::from_config_or_status(&primary_link_status)
-        .context(error::GetDnsSettingsSnafu)?;
-    Ok(dns_settings)
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "DHCPv4")]
+struct Dhcp4Section {
+    #[systemd(entry = "UseDNS")]
+    use_dns: Option<bool>,
+    #[systemd(entry = "UseDomains")]
+    use_domains: Option<bool>,
 }
 
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "DHCPv6")]
+struct Dhcp6Section {
+    #[systemd(entry = "UseDNS")]
+    use_dns: Option<bool>,
+    #[systemd(entry = "UseDomains")]
+    use_domains: Option<bool>,
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+impl InterfaceDNSDropIn {
+    /// Using API DNS settings, create an appropriate drop-in for a network interface.
+    fn from_dns_settings(settings: &DnsSettings) -> Option<Self> {
+        // If there aren't any settings, don't write a drop-in
+        if !settings.has_name_servers() && !settings.has_search_domains() {
+            return None;
+        }
+
+        // If there are name server or search domain API settings, turn off the option to get these
+        // from the DHCP server.
+        let use_dns = match settings.has_name_servers() {
+            true => Some(false),
+            false => Some(true),
+        };
+        let use_domains = match settings.has_search_domains() {
+            true => Some(false),
+            false => Some(true),
+        };
+
+        Some(Self {
+            dhcp4: Some(Dhcp4Section {
+                use_dns,
+                use_domains,
+            }),
+            dhcp6: Some(Dhcp6Section {
+                use_dns,
+                use_domains,
+            }),
+        })
+    }
+}
+
+// Gather the DNS settings if they exist, and if so create interface level drop-ins to turn off
+// DNS-via-DHCP for either name servers or search domains.  Then write a systemd-resolved
+// configuration drop-in with the global DNS settings.
+//
+// For example, if we have DNS search domain settings from the API, we don't need to get them from
+// DHCP, so the interface drop-in will turn them off and systemd-resolved will use the global
+// setting from the API which will be set in the systemd-resolved configuration file.
+#[cfg(net_backend = "systemd-networkd")]
+fn handle_dns_settings(_primary_interface: String) -> Result<()> {
+    // The DNS settings config file is always created since it is a configuration file in the API.
+    // If it is empty, don't create any interface drop-ins
+    let dns_settings = DnsSettings::from_config().context(error::GetDnsSettingsSnafu)?;
+    let interface_drop_in = match InterfaceDNSDropIn::from_dns_settings(&dns_settings) {
+        Some(d) => d,
+        None => {
+            eprintln!("No DNS configuration to write");
+            return Ok(());
+        }
+    };
```

Comment:
It seems like a mistake to not always rewrite the drop-in with up-to-date values.

This logic would be broken if it were possible to delete settings, since then we could have a drop-in written by a previous iteration.